File: EliteFramework/EliteFramework/Classes/NSLayoutConstraint/EliteLayoutConstraint.m
//
//  EliteLayoutConstraint.m
//  EliteFramework
//
//  Created by Lifelong-Study on 2016/3/22.
//  Copyright © 2016年 Lifelong-Study. All rights reserved.
//

#import "EliteLayoutConstraint.h"

@implementation NSLayoutConstraint (EliteLayoutConstraint)

//
+ (instancetype _Nonnull)constraintWithItem:(nonnull UIView *)subview attribute:(NSLayoutAttribute)attribute
{
    return [self constraintWithItem:subview
                          attribute:attribute
                          relatedBy:NSLayoutRelationEqual
                             toItem:subview.superview
                          attribute:attribute
                         multiplier:1.0
                           constant:0.0];
}

//
+ (instancetype _Nonnull)constraintWithItem:(nonnull UIView *)subview width:(CGFloat)width
{
    return [self constraintWithItem:subview
                          attribute:NSLayoutAttributeWidth
                          relatedBy:NSLayoutRelationEqual
                             toItem:nil
                          attribute:NSLayoutAttributeNotAnAttribute
                         multiplier:1.0
                           constant:width];
}

//
+ (instancetype _Nonnull)constraintWithItem:(nonnull UIView *)subview height:(CGFloat)height
{
    return [self constraintWithItem:subview
                          attribute:NSLayoutAttributeHeight
                          relatedBy:NSLayoutRelationEqual
                             toItem:nil
                          attribute:NSLayoutAttributeNotAnAttribute
                         multiplier:1.0
                           constant:height];
}

//
+ (instancetype _Nonnull)constraintEqualWidthWithItem:(nonnull UIView *)view1 toItem:(nonnull UIView *)view2
{
    return [self constraintWithItem:view1
                          attribute:NSLayoutAttributeWidth
                          relatedBy:NSLayoutRelationEqual
                             toItem:view2
                          attribute:NSLayoutAttributeWidth
                         multiplier:1.0
                           constant:0.0];
}

//
+ (instancetype _Nonnull)constraintEqualHeightWithItem:(nonnull UIView *)view1 toItem:(nonnull UIView *)view2
{
    return [self constraintWithItem:view1
                          attribute:NSLayoutAttributeHeight
                          relatedBy:NSLayoutRelationEqual
                             toItem:view2
                          attribute:NSLayoutAttributeHeight
                         multiplier:1.0
                           constant:0.0];
}

//
+ (NSArray<__kindof NSLayoutConstraint *> * _Nonnull)constraintsWithVisualFormat:(nonnull NSString *)format views:(NSDictionary<NSString *, id> * _Nonnull)views
{
    return [self constraintsWithVisualFormat:format options:NSLayoutFormatDirectionLeadingToTrailing metrics:nil views:views];
}

@end

File: EliteFramework/EliteLayoutConstraint.swift
//
//  EliteLayoutConstraint.swift
//  EliteFramework
//
//  Created by Lifelong-Study on 2016/5/28.
//  Copyright © 2016年 Lifelong-Study. All rights reserved.
//

import UIKit

public extension NSLayoutConstraint {
    
    //
    class func constraint(item item: UIView, attribute: NSLayoutAttribute) -> NSLayoutConstraint {
        return constraint(item: item, attribute: attribute, offset: 0.0)
    }
    
    //
    class func constraint(item item: UIView, attribute: NSLayoutAttribute, offset: CGFloat) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: attribute, relatedBy: .Equal, toItem: item.superview, attribute: attribute, multiplier: 1.0, constant: offset)
    }
    
    //
    class func constraints(withVisualFormat format: String, views: [String : AnyObject]) -> [NSLayoutConstraint] {
        return constraintsWithVisualFormat(format, options: .DirectionLeadingToTrailing, metrics: nil, views: views)
    }
    
    //
    class func constraints(withVisualFormat format: String, metrics: [String : AnyObject]?, views: [String : AnyObject]) -> [NSLayoutConstraint] {
        return constraintsWithVisualFormat(format, options: .DirectionLeadingToTrailing, metrics: metrics, views: views)
    }
    
    //
    class func equal(item item: UIView, toItem: UIView, attribute: NSLayoutAttribute) -> NSLayoutConstraint {
        return equal(item: item, toItem: toItem, attribute: attribute, offset: 0.0)
    }
    
    //
    class func equal(item item: UIView, toItem: UIView, attribute: NSLayoutAttribute, offset: CGFloat) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: attribute, relatedBy: .Equal, toItem: toItem, attribute: attribute, multiplier: 1.0, constant: offset)
    }
    
    //
    class func equal(item item: UIView, attribute: NSLayoutAttribute) -> NSLayoutConstraint {
        return equal(item: item, toItem: item.superview!, attribute: attribute, offset: 0.0)
    }
    
    //
    class func equal(item item: UIView, attribute: NSLayoutAttribute, offset: CGFloat) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: attribute, relatedBy: .Equal, toItem: item.superview!, attribute: attribute, multiplier: 1.0, constant: offset)
    }
    
    //
    class func equal(item item: UIView, attribute: NSLayoutAttribute, toAttribute: NSLayoutAttribute) -> NSLayoutConstraint {
        return equal(item: item, attribute: attribute, toAttribute: toAttribute, multiplier: 1.0)
    }
    
    //
    class func equal(item item: UIView, attribute: NSLayoutAttribute, toAttribute: NSLayoutAttribute, multiplier: CGFloat) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: attribute, relatedBy: .Equal, toItem: item.superview!, attribute: toAttribute, multiplier: multiplier, constant: 0.0)
    }
    
    //
    class func equal(item item: UIView, toItem: UIView) -> [NSLayoutConstraint] {
        return [equal(item: item, toItem: toItem, attribute: .Width),
                equal(item: item, toItem: toItem, attribute: .Height),
                equal(item: item, toItem: toItem, attribute: .CenterX),
                equal(item: item, toItem: toItem, attribute: .CenterY)]
    }
    
    //
    class func constraint(item item: UIView, attribute: NSLayoutAttribute, equalAttribute: NSLayoutAttribute) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: attribute, relatedBy: .Equal, toItem: item, attribute: equalAttribute, multiplier: 1.0, constant: 0.0)
    }
    
    //
    class func constraint(item item: UIView, attribute: NSLayoutAttribute, equalAttribute: NSLayoutAttribute, offset: CGFloat) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: attribute, relatedBy: .Equal, toItem: item, attribute: equalAttribute, multiplier: 1.0, constant: offset)
    }
    
    //
    class func constraint(item item: UIView, width: CGFloat) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: .Width, relatedBy: .Equal, toItem: nil, attribute: .Width, multiplier: 1.0, constant: width)
    }
    
    //
    class func constraint(item item: UIView, height: CGFloat) -> NSLayoutConstraint {
        return NSLayoutConstraint(item: item, attribute: .Height, relatedBy: .Equal, toItem: nil, attribute: .Height, multiplier: 1.0, constant: height)
    }
}
